### PR TITLE
Standardize parameter mapping APIs

### PR DESCRIPTION
**Pull request by philippjfr**
Over time we accumulated a bunch of workaround for mapping parameters to properties. Specifically we had the `_rename` dictionary used for mapping from parameter to property names, then we had various `_init_params` and `_get_properties` methods used to provide parameters to initialize and update properties on the underlying bokeh model. This PR attempts to rationalize this a bit:

- `_rename` consistently used to declare parameter mapping (i.e. avoid manual workarounds for this)
- Bokeh models are now always initialized by calling `_get_properties` which itself calls `_init_params` and then transforms those parameters with `_process_param_change`, ensuring that as much logic for mapping from Panel parameters to Bokeh properties is encapsulated in this method.
- Introduce a `ModelPane` baseclass for panes that very simply map to an underlying bokeh model and apply some transformations to parameters.

### CODE REVIEW

File: panel/pane/image.py [hunk 1]
```diff
@@ -402,26 +422,20 @@ class PDF(FileBase):
...
-        p["embed"] = self.embed
-        p["start_page"] = self.start_page
-        p["width"] = self.width
-        p["height"] = self.height
+    _rerender_params: ClassVar[List[str]] = FileBase._rerender_params + ['start_page']
 
-        if self.object is None:
-            return dict(p, text='<embed></embed>')
-        if self.embed:
+    def _transform_object(self, obj):
```
Comment (philippjfr) [inline panel/pane/image.py:429]:
```suggestion
    def _transform_object(self, obj: Any) -> Dict[str, Any]:
```

File: panel/pane/markup.py [hunk 2]
```diff
@@ -243,32 +226,33 @@ def _cleanup(self, root: Model | None = None) -> None:
...
-                          if p not in DivPaneBase.param and p != '_object'}
-                html = df.to_html(**kwargs)
+                          if p not in HTMLBasePane.param and p != '_object'}
+                html = obj.to_html(**kwargs)
         else:
             html = ''
-        return dict(properties, text=escape(html))
+        return dict(object=escape(html))
+
+    def _init_params(self):
```
Comment (philippjfr) [inline panel/pane/markup.py:248]:
```suggestion
    def _init_params(self) -> Dict[str, Any]:
```

File: panel/pane/markup.py [hunk 3]
```diff
@@ -430,13 +422,22 @@ def applies(cls, obj: Any, **params) -> float | bool | None:
...
         except Exception:
-            data = self.object
+            data = obj
         text = json.dumps(data or {}, cls=self.encoder)
-        depth = None if self.depth < 0 else self.depth
-        return dict(text=text, theme=self.theme, depth=depth,
-                    hover_preview=self.hover_preview, **properties)
+        return dict(object=text)
+
+    def _process_property_change(self, properties):
```
Comment (philippjfr) [inline panel/pane/markup.py:433]:
```suggestion
    def _process_property_change(self, properties: Dict[str, Any]) -> Dict[str, Any]:
```

File: panel/pane/markup.py [hunk 4]
```diff
@@ -430,13 +422,22 @@ def applies(cls, obj: Any, **params) -> float | bool | None:
...
+        properties = super()._process_property_change(properties)
+        if 'depth' in properties:
+            properties['depth'] = -1 if properties['depth'] is None else properties['depth']
+        return properties
+
+    def _process_param_change(self, params):
```
Comment (philippjfr) [inline panel/pane/markup.py:439]:
```suggestion
    def _process_param_change(self, params: Dict[str, Any]) -> Dict[str, Any] :
```

File: panel/reactive.py [hunk 5]
```diff
@@ -128,6 +130,25 @@ def __init__(self, **params):
...
+        return rename
+
+    @property
+    def _linked_properties(self):
+        return tuple(
+            self._property_mapping.get(p, p) for p in self.param
+            if p not in Viewable.param and self._property_mapping.get(p, p) is not None
+        )
+
+    def _get_properties(self):
```
Comment (philippjfr) [inline panel/reactive.py:149]:
```suggestion
    def _get_properties(self) -> Dict[str, Any]:
```

File: panel/reactive.py [hunk 6]
```diff
@@ -128,6 +130,25 @@ def __init__(self, **params):
     # Model API
     #----------------------------------------------------------------
 
+    @classproperty
+    @lru_cache(maxsize=None)
+    def _property_mapping(cls):
```
Comment (philippjfr) [inline panel/reactive.py:135]:
```suggestion
    def _property_mapping(cls) -> Dict[str, str | None]:
```

File: panel/widgets/select.py [hunk 7]
```diff
@@ -244,11 +247,10 @@ def _validate_options_groups(self, *events):
             )
 
     def _process_param_change(self, msg):
```
Comment (philippjfr) [inline panel/widgets/select.py:249]:
```suggestion
    def _process_param_change(self, msg: Dict[str, Any]) -> Dict[str, Any]:
```

File: panel/widgets/select.py [hunk 8]
```diff
@@ -312,6 +314,8 @@ class _MultiSelectBase(SingleSelectBase):
 
     _supports_embed: ClassVar[bool] = False
 
+    __abstract = True
+
     def _process_param_change(self, msg):
```
Comment (philippjfr) [inline panel/widgets/select.py:319]:
```suggestion
    def _process_param_change(self, msg: Dict[str, Any]) -> Dict[str, Any]:
```

**Comment by Hoxbro** [comment]
I would like to see all the type changes be made in another PR.

**Comment by philippjfr** [comment]
Fair, just to clarify that is for review purposes, to avoid merge conflicts or something else?

**Comment by Hoxbro** [comment]
Review purposes and to be able to look at the PR in the future if needed.

**Comment by philippjfr** [comment]
Done, please review in that case 😄

**Comment by philippjfr** [comment]
Want to make progress elsewhere (and merge this into the UI test PR) so I'll probably go ahead and merge. This is extremely difficult to review but if you do find the time/desire to do so I'll apply any fixes in a new PR.

File: panel/layout/base.py [hunk 9]
```diff
@@ -616,6 +625,8 @@ class ListPanel(ListLike, Panel):
         Whether to add scrollbars if the content overflows the size
         of the container.""")
 
+    _rename: ClassVar[Mapping[str, str | None]] = {'scroll': None}
```
Comment (Hoxbro) [inline panel/layout/base.py:628]: `Mapping[str, str | None]` is used a lot maybe we should make it a type alias?

File: panel/pane/alert.py [hunk 10]
```diff
@@ -30,7 +33,7 @@ class Alert(Markdown):
 
     priority: ClassVar[float | bool | None] = 0
 
-    _rename: ClassVar[Mapping[str, str | None]] = dict(Markdown._rename, alert_type=None)
+    _rename: ClassVar[Mapping[str, str | None]] = {'alert_type': None}
```
Comment (Hoxbro) [inline panel/pane/alert.py:36]: Why was `Markdown._rename` removed?

File: panel/pane/base.py [hunk 11]
```diff
@@ -1,13 +1,13 @@
 """
-Defines the PaneBase class defining the API for panes which convert
-objects to a visual representation expressed as a bokeh model.
+Defines base classes for Pane components which allow wrapping a Python
+object transforming it into a Bokeh model that can be rendered.
```
Comment (Hoxbro) [inline panel/pane/base.py:3]: Would probably update with:
`...object and transforming...`
`...that can be rendered in the browser.`

File: panel/reactive.py [hunk 12]
```diff
@@ -1651,6 +1683,7 @@ def _get_template(self) -> Tuple[str, List[str], Mapping[str, List[Tuple[str, Li
                 html = html.replace('${%s}' % child_name, '')
         return html, parser.nodes, p_attrs
 
+    @property
     def _linked_properties(self) -> List[str]:
```
Comment (Hoxbro) [inline panel/reactive.py:1687]: Should be returning tuple type.

File: panel/widgets/indicators.py [hunk 13]
```diff
@@ -274,7 +268,7 @@ class Number(ValueIndicator):
     title_size = param.String(default='18pt', doc="""
         The size of the title given by the name.""")
 
-    _rename: ClassVar[Mapping[str, str | None]] = {}
+    _rename: ClassVar[Mapping[str, str | None]] = {'name': 'name'}
```
Comment (Hoxbro) [inline panel/widgets/indicators.py:271]: I'm not sure I follow this rename... Is it because the parent class rename `name`?

**Review by Hoxbro — COMMENTED** [review]
Have added some comment.

Add some point I think we should use `list`/`tuple`/`dict` instead of  `List`/`Tuple`/`Dict`. This can be done automatic with `pyupgrade` or `ruff`.

**Pull request merged**